training/deepspeech_training/util/sample_collections.py

```diff
@@ -355,11 +409,100 @@ def samples_from_files(filenames, buffering=BUFFER_SIZE, labeled=None):
         If False: Ignores transcripts (if available) and always reads (unlabeled) util.audio.Sample instances.
         If None: Reads util.sample_collections.LabeledSample instances from sources with transcripts and
         util.audio.Sample instances from sources with no transcripts.
+
+    Returns
+    -------
+    iterable of util.sample_collections.LabeledSample (labeled=True) or util.audio.Sample (labeled=False) supporting len
     """
-    filenames = list(filenames)
-    if len(filenames) == 0:
+    sample_sources = list(sample_sources)
+    if len(sample_sources) == 0:
         raise ValueError('No files')
-    if len(filenames) == 1:
-        return samples_from_file(filenames[0], buffering=buffering, labeled=labeled)
-    cols = list(map(partial(samples_from_file, buffering=buffering, labeled=labeled), filenames))
+    if len(sample_sources) == 1:
+        return samples_from_source(sample_sources[0], buffering=buffering, labeled=labeled)
+    cols = list(map(partial(samples_from_source, buffering=buffering, labeled=labeled), sample_sources))
     return Interleaved(*cols, key=lambda s: s.duration)
+
+
+class PreparationContext:
+    def __init__(self, target_audio_type, augmentations):
+        self.target_audio_type = target_audio_type
+        self.augmentations = augmentations
+
+
+PREPARATION_CONTEXT = None
+
+
+def _init_preparation_worker(preparation_context):
+    global PREPARATION_CONTEXT  # pylint: disable=global-statement
+    PREPARATION_CONTEXT = preparation_context
```

Comment:
Why is this global context needed? Can you pass the context (or just the two values) directly to `_prepare_sample` using `functools.partial`?

Reply:
Augmentations are allowed to be stateful per process (e.g. `overlay` keeps its remainder overlay data across `apply` calls). If augmentations are applied in a multi-process scenario, the context would be passed around by copy.

Reply:
I don't understand. Using `functools.partial` would still be stateful per process. It would work exactly the same as the current code, but without using globals.

Reply:
Ah - now I got the point. Will try it.

Reply:
Unfortunately `partial` is not working in this case. This code

```Python
import random
from functools import partial
from multiprocessing.pool import Pool

class Box:
    def __init__(self):
        self.prefix = None

def prefix(n, b=None):
    if b.prefix is None:
        print('randomize prefix')
        b.prefix = random.randint(10000000, 20000000)
    return '{}: {}'.format(b.prefix, n)

pool = Pool(2)
p = partial(prefix, b=Box())
for r in pool.imap(p, range(10)):
    print(r)
```

results in the following output:

```
randomize prefix
randomize prefix
randomize prefix
randomize prefix
randomize prefix
randomize prefix
randomize prefix
randomize prefix
12323215: 0
19235836: 1
randomize prefix
randomize prefix
13304703: 2
17651007: 3
19406745: 4
17973436: 5
12521431: 6
16399146: 7
11511919: 8
18938851: 9
```

The object passed to `partial` got part of the new function's code. Therefore it is always the same (in this case carrying `None`).

Reply:
OK, sorry for the waste of time.